### PR TITLE
jsonrpc: add api to get block proof

### PR DESCRIPTION
**Pull request by karim-en**
Added json rpc api `light_client_block_proof` to get the block proof.
Inputs:
```
pub struct RpcLightClientBlockProofRequest {
    pub block_hash: near_primitives::hash::CryptoHash,
    pub light_client_head: near_primitives::hash::CryptoHash,
}
```
outputs:
```
pub struct RpcLightClientBlockProofResponse {
    pub block_header_lite: near_primitives::views::LightClientBlockLiteView,
    pub block_proof: near_primitives::merkle::MerklePath,
}
```
## Description
The current `light_client_proof` api returns two kinds of proofs:
1. Outcome proof: this proof is needed to prove that the outcome was included in a block.
2. Block proof: to prove the block is part of the chain.

Currently, we can use only the archival node to get proof of historical outcomes because the regular nodes store only blocks for the last five epochs.
To get rid of dependence on the archival node, we propose generating the proof for outcome in two steps:
1. Use the regular node to get the block proof.
2. Use any other sources to get blocks and outcome proofs. it could be [near-lake-framework](https://github.com/near/near-lake-framework-rs), [fastnear](https://github.com/fastnear/neardata-server), archival nodes or any other custom blocks indexer.

### CODE REVIEW

File: chain/jsonrpc/src/lib.rs [hunk 1]
```diff
@@ -384,6 +384,9 @@ impl JsonRpcHandler {
             "next_light_client_block" => {
                 process_method_call(request, |params| self.next_light_client_block(params)).await
             }
+            "light_client_block_proof" => {
```
Comment (khorolets) [inline chain/jsonrpc/src/lib.rs:387]:
```suggestion
            "EXPERIMENTAL_light_client_block_proof" => {
```

We have a kind of convention that all newly introduced RPC methods must have an `EXPERIMENTAL_` prefix.

After a few release cycles, if the method has proven to stay unchanged, we are ready to promote it to a regular one (w/o the prefix).

**Review by khorolets — CHANGES_REQUESTED** [review]
Wow, I am surprised we had the way to get the data you need using the `ViewClient` but didn't have a dedicated RPC method for it.

I left a change request, everything else regarding the code looks good to me 👍

I have a question regarding this:

> To get rid of dependence on the archival node, we propose generating the proof for outcome in two steps:
>
> ...
>2. Use any other sources to get blocks and outcome proofs. it could be [near-lake-framework](https://github.com/near/near-lake-framework-rs), [fastnear](https://github.com/fastnear/neardata-server), archival nodes or any other custom blocks indexer.

Did I get it correct that your proposal is related to the bridge itself, not to the `nearcore`, right?

**Comment by karim-en** [comment]
> Did I get it correct that your proposal is related to the bridge itself, not to the `nearcore`, right?

Yes, the bridge or any other service needs to get proofs for the historical data.

Also I don't think that devs cans trust the centralized indexers, so this api also will be useful to verify that the blocks and logs are really part of the chain.
But there is one downside of using this api for verification propose, it returns extra data (the proof) instead of just true or false.

**Review by bowenwang1996 — APPROVED** [review]

**Review by khorolets — APPROVED** [review]
Thank you!

File: chain/jsonrpc/src/lib.rs (hunk 1 above)
Comment (karim-en) [reply chain/jsonrpc/src/lib.rs]: Fixed in 9768ccc58795502db2d171cd5fc4f5011a3197d5

**Pull request merged**